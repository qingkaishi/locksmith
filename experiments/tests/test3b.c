#include <pthread.h>
#include <stdio.h>

int f(pthread_mutex_t *l) {
  int k, *j;
  
  pthread_mutex_init(l, NULL);
  pthread_mutex_lock(l);

  j = &k;
  if(pthread_mutex_trylock(l)) {
    *j = 5;
  }
  else {
    j = (int *) &l;
  }
  pthread_mutex_unlock(l);
  return 2;
}

void * dostuff(void * a) {
  pthread_mutex_t t;
  printf("foo");
  *(int *)a = 0;
  f(&t);
}
int main() {
  int i = 1; // not shared here
	pthread_t t1;

  pthread_create(&t1, NULL, dostuff, &i);
	i++;
}


