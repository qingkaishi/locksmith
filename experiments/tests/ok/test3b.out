************************* STARTING *************************


test3b.c:15: Warning: test3b.c:15: Probably using char* where void* is best used:
 lock(*l:test3b.c:4)*^{l:test3b.c:4}*^{&l:test3b.c:4} flows to int*^{j:test3b.c:5}.  Losing precision.


test3b.c:15: Warning: assigning incompatible pointers, conflating flow

test3b.c:15: Warning: lock is conflated! (*l:test3b.c:4)

test3b.c:15: Warning: assigning number to pointer
functions declared and used but not defined:
  printf
shared: &i:test3b.c:28

Warning: Possible data race: &i:test3b.c:28 is not protected!
 references:
  dereference of cast:test3b.c:24 at test3b.c:24
    &i:test3b.c:28 => cast:test3b.c:24 at dostuff test3b.c:31
  locks acquired:
    <empty>
  in: main at test3b.c:27 

  dereference of &i:test3b.c:28 at test3b.c:32
    &i:test3b.c:28
  locks acquired:
    <empty>
  in: main at test3b.c:27 


*************************** DONE ***************************
